homeassistant/util/async_.py

```diff
@@ -76,3 +74,36 @@ def run_callback() -> None:
 
     loop.call_soon_threadsafe(run_callback)
     return future
+
+
+async def safe_wait(
+    tasks: Any, logger: Optional[logging.Logger] = None, return_exceptions: bool = False
+) -> List[Any]:
+    """Safe version of wait and gather.
+
+    It work like gather but wait don't break any workflows.
+    It allow also to log exception in correct namespace.
+    """
+    all_tasks = list(tasks)
+    if not all_tasks:
+        return []
+    finished_tasks, _ = await asyncio.wait(all_tasks)
+
+    results: List[Any] = []
+    raise_exception: Optional[BaseException] = None
+    for task in finished_tasks:
+        if not task.done():
+            results.append(None)
+        elif task.exception():
+            if logger:
+                logger.exception(task.exception())
+            if not raise_exception:
+                raise_exception = task.exception()
```

Comment:
Won't this result in raising whatever the last exception was?  Isn't that the same behavior of `gather`? (or maybe `gather` raises the first exception?)

Would it be better to store a list of exceptions and then create a new exception type called AggregateException that takes the list of exceptions?  [Similar to this](https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception?view=netframework-4.8).

Reply:
That breaks the error handling that can follow. I would prefer to be a better gather replacement.

Reply:
I agree, AggregateException would be nice to be a feature, but without it's not work in the same way as with dot.net

Reply:
Does it really break error handling, given this PR is addressing errors that weren't being caught?  🤷‍♀ This is a breaking change anyway and introducing that exception type wouldn't be that big of a lift IMO.